internal/consensus/state.go

```diff
@@ -1311,24 +1313,33 @@ func (cs *State) enterPrevote(height int64, round int32) {
 	// (so we have more time to try and collect +2/3 prevotes for a single block)
 }
 
+func (cs *State) proposalIsTimely() bool {
+	tp := types.TimingParams{
+		Precision:    cs.state.ConsensusParams.Timing.Precision,
+		MessageDelay: cs.state.ConsensusParams.Timing.MessageDelay,
+	}
+
+	return cs.Proposal.IsTimely(cs.ProposalReceiveTime, tp, cs.state.InitialHeight)
+}
+
 func (cs *State) defaultDoPrevote(height int64, round int32) {
 	logger := cs.Logger.With("height", height, "round", round)
 
 	// We did not receive a proposal within this round. (and thus executing this from a timeout)
```

Comment:
Actually, we might have received a proposal, but for sure not the full proposed block.